File: app/src/main/java/com/ternaryop/feedly/FeedlyRateLimit.java
package com.ternaryop.feedly;

import java.io.IOException;
import java.net.HttpURLConnection;

import org.joda.time.Period;
import org.joda.time.format.PeriodFormatter;
import org.joda.time.format.PeriodFormatterBuilder;

/**
 * Created by dave on 25/02/17.
 * Hold the info about rate limits
 */

public class FeedlyRateLimit {
    public static final PeriodFormatter RESET_LIMIT_FORMAT = new PeriodFormatterBuilder()
            .appendDays()
            .appendSeparator(" days ")
            .printZeroAlways().minimumPrintedDigits(2).appendHours()
            .appendSeparator(":")
            .printZeroAlways().minimumPrintedDigits(2).appendMinutes()
            .appendSeparator(":")
            .printZeroAlways().minimumPrintedDigits(2).appendSeconds()
            .toFormatter();

    private int apiCallsCount;
    private int apiResetLimit;

    public static final FeedlyRateLimit instance = new FeedlyRateLimit();

    private FeedlyRateLimit() {
        reset();
    }

    private void reset() {
        apiCallsCount = -1;
        apiResetLimit = -1;
    }

    public void update(HttpURLConnection conn) throws IOException {
        synchronized (this) {
            reset();
            if (conn.getResponseCode() != HttpURLConnection.HTTP_OK) {
                return;
            }
            String v = conn.getHeaderField("X-Ratelimit-Count");
            if (v != null) {
                apiCallsCount = Integer.parseInt(v);
            }
            v = conn.getHeaderField("X-Ratelimit-Reset");
            if (v != null) {
                apiResetLimit = Integer.parseInt(v);
            }
        }
    }

    public int getApiCallsCount() {
        return apiCallsCount;
    }

    public int getApiResetLimit() {
        return apiResetLimit;
    }

    public String getApiResetLimitAsString() {
        if (apiResetLimit < 0) {
            return "N/A";
        }
        Period period = Period.seconds(apiResetLimit).normalizedStandard();
        return RESET_LIMIT_FORMAT.print(period);
    }
}

File: app/src/main/java/com/ternaryop/feedly/FeedlyRateLimit.kt
package com.ternaryop.feedly

import okhttp3.Headers
import java.net.HttpURLConnection

/**
 * Created by dave on 25/02/17.
 * Hold the info about rate limits
 */

@Suppress("MemberVisibilityCanBePrivate")
object FeedlyRateLimit {

    var apiCallsCount: Int = -1
        private set
    var apiResetLimit: Int = -1
        private set

    val apiResetLimitAsString: String
        get() {
            if (apiResetLimit < 0) {
                return "N/A"
            }
            return formatResetLimit(apiResetLimit)
        }

    init {
        reset()
    }

    private fun formatResetLimit(seconds: Int): String {
        val sb = StringBuilder()
        var remainingSeconds = seconds
        val days = remainingSeconds / (24 * 3600)
        remainingSeconds %= (24 * 3600)
        val hours = remainingSeconds / 3600
        remainingSeconds %= 3600
        val minutes = remainingSeconds / 60
        remainingSeconds %= 60

        if (days > 0) {
            sb.append(" $days days ")
        }
        return sb.append(String.format("%02d:%02d:%02d", hours, minutes, remainingSeconds)).toString()
    }

    private fun reset() {
        apiCallsCount = -1
        apiResetLimit = -1
    }

    fun update(responseCode: Int, headers: Headers) {
        synchronized(this) {
            reset()
            if (responseCode == HttpURLConnection.HTTP_OK) {
                apiCallsCount = headers["X-Ratelimit-Count"]?.toInt() ?: -1
                apiResetLimit = headers["X-Ratelimit-Reset"]?.toInt() ?: -1
            }
        }
    }
}
